hal/spi/spi_drv_stm32.c

```diff
@@ -25,146 +25,239 @@
+    stm_gpio_config(QSPI_IO1_PIO_BASE, QSPI_IO1_PIN, QSPI_IO1_PIN_AF, 0, 3);
+    stm_gpio_config(QSPI_IO2_PIO_BASE, QSPI_IO2_PIN, QSPI_IO2_PIN_AF, 0, 3);
+    stm_gpio_config(QSPI_IO3_PIO_BASE, QSPI_IO3_PIN, QSPI_IO3_PIN_AF, 0, 3);
+#endif
 }
 
-static void spi_tpm2_pin_setup(void)
+static void stm_pins_release(void)
 {
-#ifdef WOLFBOOT_TPM
-    uint32_t reg;
-    RCC_GPIO_CLOCK_ER |= SPI_PIO_CS_CEN;
-    reg = SPI_PIO_CS_MODE & ~ (0x03 << (SPI_CS_TPM * 2));
-    SPI_PIO_CS_MODE = reg | (1 << (SPI_CS_TPM * 2));
-    reg = SPI_PIO_CS_PUPD & ~(0x03 <<  (SPI_CS_TPM * 2));
-    SPI_PIO_CS_PUPD = reg | (0x01 << (SPI_CS_TPM * 2));
-    reg = SPI_PIO_CS_OSPD & ~(0x03 << (SPI_CS_TPM * 2));
-    SPI_PIO_CS_OSPD |= (0x03 << (SPI_CS_TPM * 2));
-    spi_cs_off(SPI_CS_TPM);
+#ifdef SPI_FLASH
+    stm_gpio_config(SPI_CLOCK_PIO_BASE, SPI_CLOCK_PIN, 0, 0, 0);
+    stm_gpio_config(SPI_MOSI_PIO_BASE, SPI_MOSI_PIN, 0, 0, 0);
+    stm_gpio_config(SPI_MISO_PIO_BASE, SPI_MISO_PIN, 0, 0, 0);
+#endif
+#ifdef QSPI_FLASH
+    stm_gpio_config(QSPI_CS_PIO_BASE, QSPI_CS_FLASH_PIN, 0, 0, 0);
+    stm_gpio_config(QSPI_CLOCK_PIO_BASE, QSPI_CLOCK_PIN, 0, 0, 0);
+    stm_gpio_config(QSPI_IO0_PIO_BASE, QSPI_IO0_PIN, 0, 0, 0);
+    stm_gpio_config(QSPI_IO1_PIO_BASE, QSPI_IO1_PIN, 0, 0, 0);
+    stm_gpio_config(QSPI_IO2_PIO_BASE, QSPI_IO2_PIN, 0, 0, 0);
+    stm_gpio_config(QSPI_IO3_PIO_BASE, QSPI_IO3_PIN, 0, 0, 0);
 #endif
 }
 
-static void RAMFUNCTION spi1_pins_setup(void)
+static void RAMFUNCTION spi_reset(void)
 {
-    uint32_t reg;
-    RCC_GPIO_CLOCK_ER |= SPI_PIO_CEN;
-    /* Set mode = AF */
-    reg = SPI_PIO_MODE & ~ (0x03 << (SPI1_CLOCK_PIN * 2));
-    SPI_PIO_MODE = reg | (2 << (SPI1_CLOCK_PIN * 2));
-    reg = SPI_PIO_MODE & ~ (0x03 << (SPI1_MOSI_PIN * 2));
-    SPI_PIO_MODE = reg | (2 << (SPI1_MOSI_PIN * 2));
-    reg = SPI_PIO_MODE & ~ (0x03 << (SPI1_MISO_PIN * 2));
-    SPI_PIO_MODE = reg | (2 << (SPI1_MISO_PIN * 2));
-
-    /* Alternate function */
-#if SPI1_CLOCK_PIN < 8
-    reg = SPI_PIO_AFL & ~(0xf << ((SPI1_CLOCK_PIN) * 4));
-    SPI_PIO_AFL = reg | (SPI1_PIN_AF << ((SPI1_CLOCK_PIN) * 4));
-#else
-    reg = SPI_PIO_AFH & ~(0xf << ((SPI1_CLOCK_PIN - 8) * 4));
-    SPI_PIO_AFH = reg | (SPI1_PIN_AF << ((SPI1_CLOCK_PIN - 8) * 4));
-#endif
-#if SPI1_MOSI_PIN < 8
-    reg = SPI_PIO_AFL & ~(0xf << ((SPI1_MOSI_PIN) * 4));
-    SPI_PIO_AFL = reg | (SPI1_PIN_AF << ((SPI1_MOSI_PIN) * 4));
-#else
-    reg = SPI_PIO_AFH & ~(0xf << ((SPI1_MOSI_PIN-8) * 4));
-    SPI_PIO_AFH = reg | (SPI1_PIN_AF << ((SPI1_MOSI_PIN - 8) * 4));
-#endif
-#if SPI1_MISO_PIN < 8
-    reg = SPI_PIO_AFL & ~(0xf << ((SPI1_MISO_PIN) * 4));
-    SPI_PIO_AFL = reg | (SPI1_PIN_AF << ((SPI1_MISO_PIN) * 4));
-#else
-    reg = SPI_PIO_AFH & ~(0xf << ((SPI1_MISO_PIN-8) * 4));
-    SPI_PIO_AFH = reg | (SPI1_PIN_AF << ((SPI1_MISO_PIN - 8) * 4));
-#endif
-
-#ifdef PLATFORM_stm32l0
-    reg = SPI_PIO_PUPD & ~(0x03 <<  (SPI1_CLOCK_PIN * 2));
```

Comment:
Please add a `PLATFORM_stm32l0` exception in `stm_pins_setup` for stm32l0 to set the appropriate pulldown and ospd during SPI pin initialization